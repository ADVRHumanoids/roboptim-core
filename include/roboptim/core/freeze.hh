// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_FREEZE_HH
# define ROBOPTIM_CORE_FREEZE_HH
# include <vector>
# include <utility>

# include <roboptim/core/fwd.hh>
# include <roboptim/core/derivable-function.hh>
# include <roboptim/core/problem.hh>

namespace roboptim
{
  /// \brief Constraint that can be used to force some arguments values.
  ///
  /// This constraint expects a vector of pairs (argument id, value) to
  /// build a function that will freeze the given arguments to their
  /// associated value.
  ///
  /// For instance, the vector: [(0, 5.), (3, -12.)] forces the first
  /// parameter to five and the fourth one to minus twelve.
  ///
  // The expression of this function is:
  /// \f[-1 + \sum_{i} (x_i - value_i)^2\f]
  /// where \f$i\f$ covers the index list specified by the vector described
  /// before, \f$x_i\f$ is i-th value of the input argument and \f$value_i\f$
  /// is this argument's fixed value.
  ///
  /// Hence, the global minimum of this function is -1 and is reached
  /// when the current position matches the frozen position.
  class Freeze : public DerivableFunction
  {
  public:
    /// \brief Pair representing an argument index and a value.
    typedef std::pair<size_t, value_type> frozenArgument_t;

    /// \brief Vector of pairs (argument index, value).
    ///
    /// This type define what are the frozen arguments and what their
    /// value.
    typedef std::vector<frozenArgument_t> frozenArguments_t;

    /// \brief Create the constraint from a vector of pairs.
    ///
    /// \param size Input size (function argument's size)
    /// \param fa Vector of pairs containing what to freeze and to what value.
    Freeze (const size_t size, const frozenArguments_t fa) throw ();

    virtual ~Freeze () throw ();

    /// \brief Add itself to an optimization problem.
    /// \param pb Problem on which the constraint will be added.
    template <typename F, typename C>
    void addToProblem (Problem<F, C>& pb) throw ()
    {
      pb.addConstraint (this, Function::makeBound (-1., -1.));
    }

  protected:
    void impl_compute (result_t& res, const argument_t& x) const throw ();
    void impl_gradient (gradient_t& grad, const argument_t& x, int i)
      const throw ();
  private:
    /// Vector which defines how to freeze values.
    const frozenArguments_t frozenArguments_;

  };
} // end of namespace roboptim.

#endif //! ROBOPTIM_FREEZE_HH
