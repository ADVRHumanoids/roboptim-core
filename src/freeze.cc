// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <roboptim/core/freeze.hh>

namespace roboptim
{
  Freeze::Freeze (const size_t size, const frozenArguments_t fa) throw ()
    : DerivableFunction (size, 1),
      frozenArguments_ (fa)
  {
    // Check that given index id are valid.
    typedef frozenArguments_t::const_iterator citer_t;
    for (citer_t it = frozenArguments_.begin (); it != frozenArguments_.end ();
	 ++it)
      assert (it->first < size);
  }

  Freeze::~Freeze () throw ()
  {
  }

  void
  Freeze::impl_compute (result_t& res, const argument_t& x) const throw ()
  {
    res (0) = 0.;
    typedef frozenArguments_t::const_iterator citer_t;
    for (citer_t it = frozenArguments_.begin (); it != frozenArguments_.end ();
	 ++it)
      res (0) += (x[it->first] - it->second) * (x[it->first] - it->second);
    res (0) -= 1;
  }

  void
  Freeze::impl_gradient (gradient_t& grad, const argument_t& x, size_type i)
    const throw ()
  {
    grad.clear ();
    typedef frozenArguments_t::const_iterator citer_t;
    for (citer_t it = frozenArguments_.begin (); it != frozenArguments_.end ();
	 ++it)
      grad (it->first) += 2 * (x[it->first] - it->second);
  }
} // end of namespace roboptim.
