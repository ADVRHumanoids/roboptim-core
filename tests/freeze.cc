// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <boost/assign/list_of.hpp>
#include <boost/numeric/ublas/io.hpp>

#include "common.hh"
#include <roboptim/core/freeze.hh>

using namespace roboptim;

int run_test ()
{
  using namespace boost::assign;

  // Freeze arguments 1, 2 and 3 (4 arguments in total).
  Freeze f (4, list_of <Freeze::frozenArgument_t> (0, 3.) (1, 7.) (2, 0.));

  // Create an argument vector.
  Freeze::vector_t v (4);
  v.clear ();

  // Evaluate at frozen point: the result is minimal (i.e. -1).
  v[0] = 3., v[1] = 7., v[2] = 0.;
  std::cout
    << "Value: " << v << std::endl
    << "Result: " << f (v) << std::endl;


  // Evaluate at another point.
  v[0] = 8., v[1] = 2., v[2] = 5., v[3] = 666.;
  std::cout
    << "Value: " << v << std::endl
    << "Result: " << f (v) << std::endl;

  return 0;
}

GENERATE_TEST ()
