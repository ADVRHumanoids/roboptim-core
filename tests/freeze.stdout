Value: [4](3,7,0,0)
Result: [1](-1)
Value: [4](8,2,5,666)
Result: [1](74)
